bundledata.go

```diff
@@ -230,6 +234,20 @@ type ApplicationSpec struct {
 	RequiresTrust bool `bson:"trust,omitempty" json:"trust,omitempty" yaml:"trust,omitempty"`
 }
 
+// Empty works out if an ApplicationSpec is empty or not.
+// We can't use an equality check here because of maps with in the type and
+// using reflect.IsZero returns false because sometimes maps can be initialized.
+func (s *ApplicationSpec) Empty() bool {
+	return s.Charm == "" && s.Channel == "" && s.Series == "" &&
+		len(s.Resources) == 0 && s.NumUnits == nil && s.Scale_ == nil &&
+		len(s.To) == 0 && s.Placement_ == "" && !s.Expose &&
+		len(s.ExposedEndpoints) == 0 && len(s.Options) == 0 &&
+		len(s.Annotations) == 0 && s.Constraints == "" &&
+		len(s.Storage) == 0 && len(s.Devices) == 0 &&
+		len(s.EndpointBindings) == 0 && len(s.Offers) == 0 &&
+		s.Plan == "" && !s.RequiresTrust
+}
```

Comment:
My Python brain wants this to be constructed something like:

```
    for key in s.keys:
        if not_empty(key):
            return True
    # if everything is empty, return empty
    return False
```

In Python, not_empty would be a function that could take anything as an arg, which branched based on type. In Erlang, not_empty would be a series of functions, pattern matched based on the type of the arg. I'm not sure what the best practice is in Go. But it definitely feels wrong to me to see a big chunk of text that needs to be hand edited if the bundle spec ever changes. (Worse, if it's not hand edited, it can return a false negative if a new key, and only the new key, is empty.)

Reply:
So you can do this automatically via the `reflect` package, but we tend to stay away from that, as it generates more complex code than you can easily reason about.

We could use `reflect.ValueOf(s).IsZero()`, but that doesn't work either, as that expects everything to be nil or the default value and that isn't true either. It could be that some fields have empty slices/maps that have been created during (de)serialization, but aren't nil, so we consider it empty if those are of length zero.

This becomes painful when you consider that this logic also has to work for overlays and then we have to ensure that every value can be part of an overlay and cause changes to the base bundle. That means that we can't just check for one value either and hopefully that'll give us enough of a sign for emptiness.

I can move back to reflect so that you can get things for free, but when somebody adds a type to the bundle and it isn't covered by all the reflect kinds (`map[interface{}]interface{}` and no `map[string]interface{}`) then you're in the same spot, but now you have the burden of understanding the empty function with the complex reflect shenanigans.

Lastly, I could use `json.Marshall` to a set of bytes and see if the length is expected, but then you need to ensure `omitempty` is in the right place. 

It's kind of annoying, but this is golang. 
Read many, write once.